### ISSUE TITLE
Customizations not working after upgrade or new install (7.11.6 & 7.11.8) 

### ISSUE THREAD

**Issue by gody01** [salesagility/SuiteCRM#7778]
After upgrading to 7.11.6 and latter 7.11.8 my Customization is not picked up any more.

#### Issue
I have changed file: 
custom/modules/EmailMan/EmailMan.php
but it is never picked up by SuiteCRM. This worked before upgrade to 7.11.6

I even reinstalled clean 7.11.8 and file is still not picked up.

#### Expected Behavior
Files in custom/moduels/ModuleName/... should override core files.

#### Actual Behavior
changes are not overrided. core file is always executed 

#### Steps to Reproduce
copy file from modules/EmailMan/EmailMan.php to  custom/modules/EmailMan/EmailMan.php 

edit copied file and for example put: 

<?php
die("HERE");

Got to new campaign and send test emails.

You should get displayed HERE on web and execution break. But it works normaly.

#### Context
All my customizations are gone :(

I think this is high priority bug.

#### Your Environment

* SuiteCRM Version used: 7.11.6, 7.11.8
* Browser name and version (e.g. Chrome Version 51.0.2704.63 (64-bit)): All
* Environment name and version (e.g. MySQL, PHP 7): mariadb-5.5.60. php-7.2.21
* Operating System and version (e.g Ubuntu 16.04): CentOS 7

**Comment by pgorod** [comment]
Not all files can be customizable with that mechanism... I wouldn't say it's a bug if one file can't be customized, we'd have to look into it, specifically.

But since you say this used to work, I am intrigued.

Is it possible that the problem happened with some other file being replaced? For example, I find examples online of this:

*In include/modules_override.php*:

```
<?php
$beanFiles['EmailMan'] = 'custom/modules/EmailMan/EmailMan.php';
```
Maybe you had this previously, and for some reason it got overwritten?

This seems to override the original values found [here](https://github.com/salesagility/SuiteCRM/blob/master/include/modules.php#L172).

But watch out, the post where I found this said it didn't work :-(

I just wonder if you had one of these schemes working for you before?

**Comment by gody01** [comment]
Hello,
Yes it definitely worked without any other file changed. But now I'm not sure if it worked on 7.11.5 or previous versions.

I will try to revert to older versions on my test setup.

It may be also problem with OS switch and consequently PHP version switch. We used to run SuiteCRM on openSuSE, now we run it on CentOS.

Does anyhow can point me to code, where decision which file is loaded is made ? I cant find it in source.

**Comment by pgorod** [comment]
It's hard to find the call because it's calling or requiring through a variable name.

Maybe [here](https://github.com/salesagility/SuiteCRM/blob/master/modules/Emails/Email.php#L1014) or [here](https://github.com/salesagility/SuiteCRM/blob/master/modules/Emails/Email.php#L1414)

But those are just guesses. One way to find this out is set a breakpoint there, and examine the call stack. Or without a debugger - provoke a fatal error there, and turn on debug backtraces so that the stack is output onto the logs.

**Comment by gody01** [comment]
I'm doing latter, yes :)

**Comment by gody01** [comment]
It seems, that chaning all require_once with require_once(get_custom_file_if_exists ...
should make all files customizable.
If there would be some interest for this I can do this over weekend or during next week and prepare huge PR ...

**Comment by pgorod** [comment]
I've done a few PR's like that. But I also did some that weren't accepted, and with good reason.

Basically, the "replace file in custom" is the most basic customizations mechanism. It is also the easiest to use, so people like it, but...

- The Extension mechanism allows for the same thing, but aggregates different contributions into coherent packs that SuiteCRM can use. This way, different developers, different features, different add-ons, can share the same customization opportunities without "stepping on each others' toes".

- then there are Class extensions. This is the more technical, but also more powerful and clean mechanism. Some places you can declare child classes that override the methods they need to override. A good example is extending Views.

So you need to take this into consideration if you decide to make a PR to add `get_custom_file_if_exists` somewhere. If it already allows a more evolved mechanism, then we shouldn't encourage people to use the more basic mechanism. But if not, sure.

I don't know if that was clear, maybe not....

**Comment by gody01** [comment]
Yes, I know for Extension Framework (although some examples on Doc site would be appreciated:) )

However for small local changes this basic customization is really handy and it would be neat, if it would "just work". And it seems, that replacing require_once, include_once. require and include to use get_custom_file_if_exists would just do it.

But PR would be very large, as those require's are in almost all files ...

**Comment by pgorod** [comment]
I would agree with such a PR, but _only_ for the cases where the better mechanisms are really not available. If they are, we shouldn't encourage people to use worse options. 

It's "quick and dirty" when you just want to get something done, but when you need to maintain systems with undocumented customizations it can really make things worse...

**Comment by Dillon-Brown** [comment]
Can raise this as a suggestion for now, I was unable to get this working on any version of 7.11.x without the file change to ```modules_override.php```.
